File: src/main/java/custom/tibame201020/adbOpenCv/script/gearScript/gear/RedGearUpgradeAdapter.java
package custom.tibame201020.adbOpenCv.script.gearScript.gear;

import java.util.ArrayList;
import java.util.List;

public class RedGearUpgradeAdapter implements GearUpgradeAdapter {

    @Override
    public boolean isOkToUpgrade(GearDTOs.Gear gear) {
        GearDTOs.GearSet gearSet = gear.metadata().set();
        GearDTOs.GearType gearType = gear.metadata().type();
        GearDTOs.GearProp gearProp = gear.prop();
        GearDTOs.GearMainProp gearMainProp = gear.metadata().mainProp();
        int gearLevel = gear.metadata().level();

        var speed = gear.prop().speed();

        if (!gear.metadata().type().equals(GearDTOs.GearType.SHOES)) {
            if ((gearLevel >= 0 && gearLevel < 9) && speed >= 5) {
                System.err.println("--- 速度足夠: " + speed);
                return true;
            }

            if ((gearLevel >= 9 && gearLevel < 12) && speed >= 10) {
                System.err.println("--- 速度足夠: " + speed);
                return true;
            }

            if ((gearLevel >= 12) && speed >= 15) {
                System.err.println("--- 速度足夠: " + speed);
                return true;
            }
        }


        boolean mainPropRequired = mainPropRequired(gearSet, gearType, gearMainProp);
        if (!mainPropRequired) {
            System.err.println("--- 裝備主屬性不符合: " + gearMainProp + ", 部位: " + gearType + ", 套裝: " + gearSet);
            return false;
        }
        boolean gearPropNeeded = gearPropNeeded(gearSet, gearProp, gearMainProp);
        if (!gearPropNeeded) {
            System.err.println("--- 裝備屬性不符合需求");
            return false;
        }

        int score = gear.metadata().score();
        var calcScore = gearProp.reduceMainProp(gearMainProp).calcScore();

        System.err.println("score: " + score + ", calcScore: " + calcScore);

        if (gearLevel >= 0 && gearLevel < 3) {
            var result = score >= 22;
            if (!result) {
                System.err.println("--- 裝備評分不足: " + score + ", 強化等級: " + gearLevel);
            }
            return result;
        }

        if (gearLevel >= 3 && gearLevel < 6) {
            var result = score >= 32;
            if (!result) {
                System.err.println("--- 裝備評分不足: " + score + ", 強化等級: " + gearLevel);
            }
            return result;
        }

        if (gearLevel >= 6 && gearLevel < 9) {
            var result = score >= 45;
            if (!result) {
                System.err.println("--- 裝備評分不足: " + score + ", 強化等級: " + gearLevel);
            }
            return result;
        }

        if (gearLevel >= 9 && gearLevel < 12) {
            var result = score >= 60;
            if (!result) {
                System.err.println("--- 裝備評分不足: " + score + ", 強化等級: " + gearLevel);
            }
            return result;
        }

        if (gearLevel >= 12) {
            var result = score >= 70;
            if (!result) {
                System.err.println("--- 裝備評分不足: " + score + ", 強化等級: " + gearLevel);
            }
            return result;
        }

        return score >= 85;
    }


    @Override
    public boolean isOkToStore(GearDTOs.Gear gear) {
        GearDTOs.GearSet gearSet = gear.metadata().set();
        GearDTOs.GearType gearType = gear.metadata().type();
        GearDTOs.GearProp gearProp = gear.prop();
        GearDTOs.GearMainProp gearMainProp = gear.metadata().mainProp();

        var speed = gear.prop().speed();
        if (speed >= 20 && !gear.metadata().type().equals(GearDTOs.GearType.SHOES)) {
            System.err.println("--- 速度足夠");
            return true;
        }

        boolean mainPropRequired = mainPropRequired(gearSet, gearType, gearMainProp);
        if (!mainPropRequired) {
            return false;
        }
        boolean gearPropNeeded = gearPropNeeded(gearSet, gearProp, gearMainProp);
        if (!gearPropNeeded) {
            return false;
        }

        int score = gear.metadata().score();
        var calcScore = gearProp.reduceMainProp(gearMainProp).calcScore();
        System.err.println("score: " + score + ", calcScore: " + calcScore);

        return score >= 85;
    }

    boolean mainPropRequired(GearDTOs.GearSet gearSet, GearDTOs.GearType gearType, GearDTOs.GearMainProp mainProp) {
        if (gearType.equals(GearDTOs.GearType.WEAPON) || gearType.equals(GearDTOs.GearType.HELMET) || gearType.equals(GearDTOs.GearType.ARMOR)) {
            return true;
        }

        if (gearType.equals(GearDTOs.GearType.SHOES) && gearSet.equals(GearDTOs.GearSet.SPEED)) {
            return mainProp.equals(GearDTOs.GearMainProp.SPEED);
        }

        if (gearType.equals(GearDTOs.GearType.NECKLACE) || gearType.equals(GearDTOs.GearType.RING) || gearType.equals(GearDTOs.GearType.SHOES)) {
            return !(mainProp.equals(GearDTOs.GearMainProp.ATK_FLAT) ||
                    mainProp.equals(GearDTOs.GearMainProp.LIFE_FLAT) ||
                    mainProp.equals(GearDTOs.GearMainProp.DEF_FLAT));

        }

        return false;
    }

    boolean gearPropNeeded(GearDTOs.GearSet gearSet, GearDTOs.GearProp gearProp, GearDTOs.GearMainProp mainProp) {
        List<GearDTOs.GearPropBelong> belongs = detectGearBelong(gearProp, mainProp);
        System.err.println("裝備適性: " + belongs);

        if (belongs.isEmpty()) {
            return false;
        }

        return switch (gearSet) {
            case ATTACK, RAGE, TORRENT, LIFE_STEAL, DUAL_ATTACK -> {
                System.err.println("打手 套裝: " + gearSet);
                yield belongs.contains(GearDTOs.GearPropBelong.DAMAGE);
            }
            case HEALTH, DEFENSE -> {
                System.err.println("坦打/坦克/輔助 套裝: " + gearSet);
                yield !belongs.contains(GearDTOs.GearPropBelong.DAMAGE);
            }
            case DESTRUCTION, CRITICAL, COUNTER, PENETRATION, INJURY ->
            {
                System.err.println("打手/坦打 套裝: " + gearSet);
                yield belongs.contains(GearDTOs.GearPropBelong.DAMAGE) || belongs.contains(GearDTOs.GearPropBelong.TANK_DAMAGE);
            }
            case PROTECTION -> {
                System.err.println("坦克 套裝: " + gearSet);
                yield belongs.contains(GearDTOs.GearPropBelong.TANK);
            }
            case HIT, RESISTANCE, SPEED, REVENGE, IMMUNITY -> {
                System.err.println("通用 套裝: " + gearSet);
                yield true;
            }
        };
    }

    List<GearDTOs.GearPropBelong> detectGearBelong(GearDTOs.GearProp gearProp, GearDTOs.GearMainProp mainProp) {
        List<GearDTOs.GearPropBelong> belongs = new ArrayList<>();
        if (GearDTOs.GearPropBelong.DAMAGE.isGearPropBelong(gearProp.reduceMainProp(mainProp))) {
            belongs.add(GearDTOs.GearPropBelong.DAMAGE);
        }
        if (GearDTOs.GearPropBelong.TANK_DAMAGE.isGearPropBelong(gearProp.reduceMainProp(mainProp))) {
            belongs.add(GearDTOs.GearPropBelong.TANK_DAMAGE);
        }
        if (GearDTOs.GearPropBelong.TANK.isGearPropBelong(gearProp.reduceMainProp(mainProp))) {
            belongs.add(GearDTOs.GearPropBelong.TANK);
        }
        if (GearDTOs.GearPropBelong.SUPPORT.isGearPropBelong(gearProp.reduceMainProp(mainProp))) {
            belongs.add(GearDTOs.GearPropBelong.SUPPORT);
        }
        return belongs;
    }
}

File: core/script/gearScript/gear/red_gear_upgrade_adapter.py
"""
紅裝升級適配器
"""
from typing import List
from .gear_upgrade_adapter import GearUpgradeAdapter
from .gear_dtos import (
    Gear, GearSet, GearType, GearMainProp, GearProp, GearPropBelong
)


class RedGearUpgradeAdapter(GearUpgradeAdapter):
    """紅裝升級適配器"""
    
    def is_ok_to_upgrade(self, gear: Gear) -> bool:
        """判斷紅裝是否可以升級"""
        gear_set = gear.metadata.set
        gear_type = gear.metadata.type
        gear_prop = gear.prop
        gear_main_prop = gear.metadata.main_prop
        gear_level = gear.metadata.level
        
        speed = gear.prop.speed
        
        if gear_type != GearType.SHOES:
            if (0 <= gear_level < 9) and speed >= 5:
                print(f"--- 速度足夠: {speed}, 強化等級: {gear_level}, 需求: 5")
                return True
            
            if (9 <= gear_level < 12) and speed >= 10:
                print(f"--- 速度足夠: {speed}, 強化等級: {gear_level}, 需求: 10")
                return True
            
            if gear_level >= 12 and speed >= 15:
                print(f"--- 速度足夠: {speed}, 強化等級: {gear_level}, 需求: 15")
                return True
        
        main_prop_required = self.main_prop_required(gear_set, gear_type, gear_main_prop)
        if not main_prop_required:
            print(f"--- 裝備主屬性不符合: {gear_main_prop.value}, 部位: {gear_type.value}, 套裝: {gear_set.value}")
            return False
        
        gear_prop_needed = self.gear_prop_needed(gear_set, gear_prop, gear_main_prop)
        if not gear_prop_needed:
            print("--- 裝備屬性不符合需求")
            return False
        
        score = gear.metadata.score
        calc_score = gear_prop.reduce_main_prop(gear_main_prop).calc_score()
        
        print(f"score: {score}, calcScore: {calc_score}")
        
        if 0 <= gear_level < 3:
            print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 20")
            return score >= 20
        elif 3 <= gear_level < 6:
            print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 30")
            return score >= 30
        elif 6 <= gear_level < 9:
            print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 40")
            return score >= 40
        elif 9 <= gear_level < 12:
            print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 55")
            return score >= 55
        elif gear_level >= 12:
            print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 65")
            return score >= 65
        
        print(f"--- 裝備評分: {score}, 強化等級: {gear_level}, 評分需求: 80 || calcScore: 70")
        return score >= 80 or calc_score >= 70
    
    def is_ok_to_store(self, gear: Gear) -> bool:
        """判斷紅裝是否可以儲存"""
        return self.is_ok_to_upgrade(gear)
    
    def main_prop_required(self, gear_set: GearSet, gear_type: GearType, main_prop: GearMainProp) -> bool:
        """判斷主屬性是否符合需求"""
        if gear_type in [GearType.WEAPON, GearType.HELMET, GearType.ARMOR]:
            return True
        
        if gear_type == GearType.SHOES and gear_set == GearSet.SPEED:
            return main_prop == GearMainProp.SPEED
        
        if gear_type in [GearType.NECKLACE, GearType.RING, GearType.SHOES]:
            return main_prop not in [GearMainProp.ATK_FLAT, GearMainProp.LIFE_FLAT, GearMainProp.DEF_FLAT]
        
        return False
    
    def gear_prop_needed(self, gear_set: GearSet, gear_prop: GearProp, main_prop: GearMainProp) -> bool:
        """判斷裝備屬性是否符合需求"""
        belongs = self.detect_gear_belong(gear_prop, main_prop)
        belong_texts = [b.value for b in belongs]
        print(f"裝備適性: {belong_texts}")
        
        if not belongs:
            return False
        
        if gear_set in [GearSet.ATTACK, GearSet.RAGE, GearSet.TORRENT, GearSet.LIFE_STEAL, GearSet.DUAL_ATTACK]:
            print(f"{gear_set.value} -> 打手套裝")
            return GearPropBelong.DAMAGE in belongs
        
        if gear_set in [GearSet.HEALTH, GearSet.DEFENSE]:
            print(f"{gear_set.value} -> 坦打/坦克/輔助套裝")
            return (GearPropBelong.SUPPORT in belongs or 
                   GearPropBelong.TANK in belongs or 
                   GearPropBelong.TANK_DAMAGE in belongs)
        
        if gear_set in [GearSet.DESTRUCTION, GearSet.CRITICAL, GearSet.COUNTER, 
                       GearSet.PENETRATION, GearSet.INJURY]:
            print(f"{gear_set.value} -> 打手/坦打套裝")
            return (GearPropBelong.DAMAGE in belongs or 
                   GearPropBelong.TANK_DAMAGE in belongs)
        
        if gear_set == GearSet.PROTECTION:
            print(f"{gear_set.value} -> 坦克套裝")
            return GearPropBelong.TANK in belongs
        
        if gear_set in [GearSet.HIT, GearSet.RESISTANCE, GearSet.SPEED, 
                       GearSet.REVENGE, GearSet.IMMUNITY]:
            print(f"{gear_set.value} -> 通用套裝")
            return True
        
        return False
    
    def detect_gear_belong(self, gear_prop: GearProp, main_prop: GearMainProp) -> List[GearPropBelong]:
        """偵測裝備歸屬"""
        belongs = []
        reduced_prop = gear_prop.reduce_main_prop(main_prop)
        
        if GearPropBelong.DAMAGE.is_gear_prop_belong(reduced_prop):
            belongs.append(GearPropBelong.DAMAGE)
        if GearPropBelong.TANK_DAMAGE.is_gear_prop_belong(reduced_prop):
            belongs.append(GearPropBelong.TANK_DAMAGE)
        if GearPropBelong.TANK.is_gear_prop_belong(reduced_prop):
            belongs.append(GearPropBelong.TANK)
        if GearPropBelong.SUPPORT.is_gear_prop_belong(reduced_prop):
            belongs.append(GearPropBelong.SUPPORT)
        
        return belongs









